hivelink-core/src/main/java/org/apache/iceberg/hivelink/core/HiveMetadataPreservingTableOperations.java

```diff
@@ -209,6 +221,67 @@ protected void doCommit(TableMetadata base, TableMetadata metadata) {
     }
   }
 
+  /**
+   * [LINKEDIN] Due to an issue that the table read in is sometimes corrupted and has incorrect columns, compare the
+   * table columns to the avro.schema.literal property (if it exists) and fix the table columns if there is a mismatch
+   */
+  static void fixMismatchedSchema(Table table) {
+    String avroSchemaLiteral = getAvroSchemaLiteral(table);
+    if (Strings.isNullOrEmpty(avroSchemaLiteral)) {
+      return;
+    }
+    Schema schema = new Schema.Parser().parse(avroSchemaLiteral);
+    List<FieldSchema> hiveCols;
+    try {
+      hiveCols = getColsFromAvroSchema(schema);
+    } catch (SerDeException e) {
+      LOG.error("Failed to get get columns from avro schema when checking schema", e);
+      return;
+    }
+
+    boolean schemaMismatched;
+    if (table.getSd().getCols().size() != hiveCols.size()) {
+      schemaMismatched = true;
+    } else {
+      Map<String, String> hiveFieldMap = hiveCols.stream().collect(
+          Collectors.toMap(FieldSchema::getName, FieldSchema::getType));
+      Map<String, String> tableFieldMap = table.getSd().getCols().stream().collect(
+          Collectors.toMap(FieldSchema::getName, FieldSchema::getType));
+      schemaMismatched = !hiveFieldMap.equals(tableFieldMap);
+    }
+
+    if (schemaMismatched) {
+      LOG.warn("Schema columns don't match avro.schema.literal, setting columns to avro.schema.literal. Schema " +
+              "columns: {}, avro.schema.literal columns: {}",
+          table.getSd().getCols().stream().map(Object::toString).collect(Collectors.joining(", ")),
+          hiveCols.stream().map(Object::toString).collect(Collectors.joining(", ")));
+      table.getSd().setCols(hiveCols);
+    }
+  }
+
+  static List<FieldSchema> getColsFromAvroSchema(Schema schema)
+      throws SerDeException {
+    AvroObjectInspectorGenerator avroOI = new AvroObjectInspectorGenerator(schema);
+    List<String> columnNames = avroOI.getColumnNames();
+    List<TypeInfo> columnTypes = avroOI.getColumnTypes();
+    if (columnNames.size() != columnTypes.size()) {
+      throw new IllegalStateException();
+    }
+
+    return IntStream.range(0, columnNames.size())
+        .mapToObj(i -> new FieldSchema(columnNames.get(i), columnTypes.get(i).getTypeName(), ""))
```

Comment:
Are we sure TypeInfo's type name representation can be used to construct `FieldSchema` ? In other words, are they using the same type name representation, even for nested types?

Reply:
Yes for nested types the type will be for example `struct<field1:string,field2:string>`. I updated the unit test to fix a mismatch on a nested field.